# Alignment to a reference transcriptome with BWA

The program BWA uses three main algorithms to perform alignments. SAMtools is a program used for processing the alignments. Download [BWA][1] and [SAMtools][2] before amending/running the following script:

```
#!/bin/bash
#module load bwa/0.6.2
#module load samtools/1.2

cut -f1 -d " " ../Directory_with_reference_FASTA_file > bwa_ref.fasta

bwa index -a is bwa_ref.fasta \
> bwa_index.stdout 2> bwa_index.stderr

for file in /Directory_with_sickle_ouput_files/*R1_qt.fastq
do
echo $file
sample=`echo $file | cut -f1 -d "_" | cut -f6 -d "/"` 
echo $sample

bwa mem -M -t 16 bwa_ref.fasta "$sample"_R1_qt.fastq "$sample"_R2_qt.fastq \
> "$sample"_mem.sam
2> "$sample"_PE_bwa_mem.stderr &

done
```
Let's take a closer look at what each part of this script is doing.

The ```bwa index``` command is used to set your reference transcriptome

```
cut -f1 -d " " ../Directory_with_reference_FASTA_file > bwa_ref.fasta

bwa index -a is bwa_ref.fasta \
> bwa_index.stdout 2> bwa_index.stderr

```
The for loop in this second part of the script will align the sequences. Note that ```bwa mem``` requires that the forward and reverse reads be together when running in batches because. The user tells BWA who the pairs are and then the output goes straight to the .sam file, skipping the .sai step of ```bwa aln```.

```
for file in /Directory_with_sickle_ouput_files/*R1_qt.fastq
do
echo $file
sample=`echo $file | cut -f1 -d "_" | cut -f6 -d "/"` 
echo $sample

bwa mem -M -t 16 bwa_ref.fasta "$sample"_R1_qt.fastq "$sample"_R2_qt.fastq \
> "$sample"_mem.sam
2> "$sample"_PE_bwa_mem.stderr &

done
```
The next step in the workflow is [Processing alignments and creating a table of raw counts] [3]  

[1]:https://github.com/lh3/bwa  
[2]:https://github.com/samtools/samtools  
[3]:https://github.com/lkomoro/RNA-Seq-Workflow/blob/master/processing_alignments.Rmd  