##Processing alignments and creating a table of raw counts

The alignment outputs must be processed and reformatted in order to be compatible with the differential expression analysis programs.

###Convert .sam files to .bam files

The .sam files outputted by the alignment software must first be converted to .bam files. Do this separately for single-end and paired-end reads.

Run the following script:

```
for file in /Directory_with_all_of_your_sam_files/*.sam
do
echo $file
sample=`echo $file | cut -f1 -d "_" | cut -f6 -d "/"` 
echo $sample

/usr/local/samtools-1.2/bin/samtools view -bh -@6 -F2048 "$sample"_mem.sam | \
/usr/local/samtools-1.2/bin/samtools sort -m 16G -@6 -O bam \
-T temporarysort -o "$sample"_PEsort.bam - \
> "$sample"_PE_sort.stdout 2> "$sample"_PE_sort.stderr
done

```
Note: If you only want to keep proper pairs, use the –f2 option in this script.

###Index the .bam file and generate a raw counts table for each sample

Run the following script:

```
for file in /Directory_with_all_of_your_bam_files/*.bam
do
echo $file
sample=`echo $file | cut -f1 -d "_" | cut -f6 -d "/"` 
echo $sample

/usr/local/samtools-1.3/bin/samtools idxstats "$sample"_PEsort.bam \
> "$sample"_PEsort.bam.stdout 2> "$sample"_PEsort.bam.bai.stderr \

/usr/local/samtools-1.3/bin/samtools idxstats "$sample"_singleton_sort.bam \
> "$sample"_singletonsort.bam.stdout 2> "$sample"_singletonsort.bam.stderr \

done
```

The outputted table should look something like this:

![](https://github.com/lkomoro/RNA-Seq-Workflow/blob/master/index_table.jpg)

####If there are unmapped reads, it may be helpful to investigate the samflags. Use the following code:

```
samtools view “bamfilename”|grep “gene/transcriptID”|less

```
The number in the second column is the samflag. Next, go to [this website][1] and type in the samflag so that the website can tell you what it means.

[1]: https://broadinstitute.github.io/picard/explain-flags.html

#### At this point, it is optional but encouraged to run samtools flagstats to get a summary of aligned reads

To obtain the aligned read summary, run the following script:

```
for file in /Directory_with_all_of_your_bam_files/*.bam
do
echo $file
sample=`echo $file | cut -f1 -d "_" | cut -f6 -d "/"` 
echo $sample

/usr/local/samtools-1.3/bin/samtools flagstat "$sample"_PEsort.bam \
> "$sample"_PEsort.bam.flagstat 2> "$sample"_PEsort.bam.bai.stderr \

/usr/local/samtools-1.3/bin/samtools flagstat "$sample"_singleton_sort.bam \
> "$sample"_singletonsort.bam.flagstat 2> "$sample"_singletonsort.bam.stderr \

done
```

###Combine the raw counts tables and convert to a matrix

The single-end and paired-end count tables for each sample must first be combined into a new file. Run the following R script to do this:

```
# combining tables
#setwd("Your_Directory")
library(plyr)
#read in data
file.list <- list.files(pattern = "counttable.txt")
sample_ID<-strsplit(file.list, '_')
f <- function(x) x[[1]]
t1 <- unlist(lapply(sample_ID, f))
#t <- unlist(sample_ID)
#maxval<-length(file.list)
#t1 <- t[seq(1, maxval, by=2)]
#t1 <- t[c(1, 3, 5, 7)]  
t2 <- unique(t1)  

for (i in 1:length(t2)){
  idx <- grep(t2[i], file.list)
  file1 <- file.list[idx[1]]
  file2 <- file.list[idx[2]]
print(file1)
print(file2)#this goes to nohup.out (if specified in command line),
  #just as a double check that the right files are being read in together
  data_PE<-read.table(file.list[idx[1]], header=FALSE)
  data_S<-read.table(file.list[idx[2]], header=FALSE)

#rename columns
data_PE<-rename(data_PE,c("V1"="GeneID_TranscriptID","V2"="Length_PE","V3"="Mapped_Counts_PE","V4"="Unmapped_Counts_PE"))
data_S<-rename(data_S,c("V1"="GeneID_TranscriptID","V2"="Length_S","V3"="Mapped_Counts_S","V4"="Unmapped_Counts_S"))

#Merge PE with singleton files, get rid of extra columns
data_comb<-merge(data_PE,data_S, all=TRUE)
data_comb_short<-data_comb[,c(1,3,6)]

#Add together,get rid of individual columns:
data_comb_short$Mapped_Counts_PE_S<-data_comb_short$Mapped_Counts_PE+data_comb_short$Mapped_Counts_S
data_comb_short1<-data_comb_short[,c(1,4)]

#write new file
write.csv(data_comb_short1, file = paste(strsplit(file.list[idx[1]], "_")[[1]][1],"PEandS_rawcounts", ".csv", sep = ""))
#can change this to be a write .txt if prefer, send it to its own new directory, etc.

#generate some summary stats of contribution of PE vs singletons
PE_mapped<-sum(data_comb_short$Mapped_Counts_PE)
S_mapped<-sum(data_comb_short$Mapped_Counts_S)
Total_mapped<-sum(data_comb_short$Mapped_Counts_PE_S)
#str(data_comb_short)
Proportion_S_mapped_of_total<-S_mapped/Total_mapped
Sample.name<-strsplit(file.list[idx[1]], "_")[[1]][1]
sum_stats<-cbind(Sample.name,PE_mapped, S_mapped, Total_mapped, Proportion_S_mapped_of_total)
write.table(sum_stats,file="mapping_sum_stats.txt",append=TRUE,sep="\t",col.names = FALSE,quote=FALSE)#hacky way to just get a table summarizing all the coverage of PE and S across samples}
```

The combined counts table must then be converted into a matrix. Run the following R script:

```
rm(list=ls())
#setwd("Your_directory
library(plyr)
file.list <- list.files(pattern = "_rawcounts.csv")

base<-read.csv(file.list[1])
base1<-base[,2]
base1<-data.frame(base1)
base1<-rename(base1,c("base1"="GeneID_TranscriptID")) 
for (i in file.list) {
  sample<-read.csv(i)
  new.name <- paste(strsplit(i, "PE")[[1]][1])
sample<-sample[,2:3]
sample<-rename(sample,c("GeneID_TranscriptID"="GeneID_TranscriptID","Mapped_Counts_PE_S"=new.name))
base1<-merge(base1,sample,by="GeneID_TranscriptID")
}

write.table(base1,file="filename",sep="\t",quote=FALSE)
```
###Roll up  the counts from transcripts to genes

Run the following R script

```
rm(list=ls())#just in case, start fresh
#setwd("Your_Directory") 
getwd()
library(reshape2)
X<-read.table("filename_conbined_rawcounts.txt")
X<-X[2:”your number”,] # this just removes the top row with *=counts of unmapped reads
X_key_table<-read.table("file_isotigs_plus_trinity_assembly_correspondence_table.txt", header=TRUE)
X_comb<-merge(X,X_key_table, by="GeneID_TranscriptID", all.x=TRUE)

length(unique(X_comb$GeneID))#how many genes are there?
samples<-colnames(X_comb, do.NULL = TRUE)
samples<-samples[c(“your numbers”)]
print(samples)

#aggregate and write to file####
filename_RNASeq_by_genes<-aggregate(cbind(“list of Gene IDs”) ~ GeneID, X_comb, sum)

write.table(filename_RNASeq_by_genes,file="filename_RNASeq_rolledupto_genes.txt",sep="\t",quote=FALSE)
```

The next step in the workflow is to perform differential expression analysis. 
